hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SnapshotChainManager.java

```diff
@@ -0,0 +1,446 @@
+import java.util.LinkedHashMap;
+import java.util.TreeMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used for creating and accessing Snapshot Chains.
+ *
+ * The snapshot chain maintains the in-memory sequence of snapshots
+ * created in chronological order.  There are two such snapshots maintained
+ * i.) Path based snapshot chain, sequence of snapshots created for a
+ * given /volume/bucket
+ * ii.) Global snapshot chain, sequence of all snapshots created in order
+ *
+ * On start, the snapshot chains are initialized from the on disk
+ * SnapshotInfoTable from the om RocksDB.
+ */
+public class SnapshotChainManager {
+  private LinkedHashMap<String, SnapshotChainInfo>  snapshotChainGlobal;
+  private Map<String, LinkedHashMap<String, SnapshotChainInfo>>
+      snapshotChainPath;
+  private Map<String, String> latestPathSnapshotID;
+  private String latestGlobalSnapshotID;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SnapshotChainManager.class);
+
+  public SnapshotChainManager(OMMetadataManager metadataManager)
+          throws IOException {
+    snapshotChainGlobal = new LinkedHashMap<>();
+    snapshotChainPath = new HashMap<>();
+    latestPathSnapshotID = new HashMap<>();
+    latestGlobalSnapshotID = null;
+    loadFromSnapshotInfoTable(metadataManager);
+  }
+
+  private void addSnapshotGlobal(String snapshotID,
+                            String prevGlobalID) throws IOException {
+    // set previous snapshotID to null if it is "" for
+    // internal in-mem structure
+    if (prevGlobalID != null && prevGlobalID.equals("")) {
+      prevGlobalID = null;
+    }
+    // on add snapshot; set previous snapshot entry nextSnapshotID =
+    // snapshotID
+    if (prevGlobalID != null &&
+        snapshotChainGlobal.containsKey(prevGlobalID)) {
+      snapshotChainGlobal
+          .get(prevGlobalID)
+          .setNextSnapshotID(snapshotID);
+    }
+    if (prevGlobalID != null &&
+        !snapshotChainGlobal.containsKey(prevGlobalID)) {
+      throw new IOException("Snapshot Chain corruption: "
+          + " previous snapshotID given but no associated snapshot "
+      + "found in snapshot chain: SnapshotID "
+      + prevGlobalID);
+    }
+    snapshotChainGlobal.put(snapshotID,
+            new SnapshotChainInfo(snapshotID, prevGlobalID, null));
+
+    // set state variable latestGlobal snapshot entry to this snapshotID
+    latestGlobalSnapshotID = snapshotID;
+  };
+
+  private void addSnapshotPath(String snapshotPath,
+                               String snapshotID,
+                                String prevPathID) throws IOException {
+    // set previous snapshotID to null if it is "" for
+    // internal in-mem structure
+    if (prevPathID != null && prevPathID.equals("")) {
+      prevPathID = null;
+    }
+
+    // on add snapshot; set previous snapshot entry nextSnapshotID =
+    // snapshotID
+    if (prevPathID != null &&
```

Comment:
Thanks @sadanand48 for your comment.  We are ok here with the snapshotPath doesn't exist.  The in is ORed expression, the first of the two will fail since snapshotPath does not exist and the second will not be executed (no NPE).
`((!snapshotChainPath .containsKey(snapshotPath)) || (!snapshotChainPath .get(snapshotPath) .containsKey(prevPathID))`
Here the first is true and so the second is not executed (no NPE).

Reply:
Thanks for explaining @neils-dev , got it.